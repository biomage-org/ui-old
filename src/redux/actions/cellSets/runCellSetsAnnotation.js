import {
  CELL_SETS_ERROR, CELL_SETS_CLUSTERING_UPDATING,
} from 'redux/actionTypes/cellSets';

import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
import fetchWork from 'utils/work/fetchWork';

const runCellSetsAnnotation = (experimentId, species, tissue) => async (dispatch, getState) => {
  const { error, updatingClustering, loading } = getState().cellSets;

  if ((loading && updatingClustering) || error) return;

  const body = {
    name: 'ScTypeAnnotate',
    species,
    tissue,
  };

  dispatch({
    type: CELL_SETS_CLUSTERING_UPDATING,
  });

  const timeout = getTimeoutForWorkerTask(getState(), 'ClusterCells');

  try {
    await fetchWork(
      experimentId,
      body,
      getState,
      dispatch,
      {
        timeout,
        broadcast: true,
      },
    );
  } catch (e) {
    dispatch({
      type: CELL_SETS_ERROR,
      payload: {
        experimentId,
        error: e,
      },
    });
  }
};

export default runCellSetsAnnotation;
