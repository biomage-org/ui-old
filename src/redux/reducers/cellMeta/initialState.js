const mitochondrialContent = {
  loading: true,
  error: false,
  data: [],
};

const doubletScores = {
  loading: true,
  error: false,
  data: [],
};

const numOfGenes = {
  loading: true,
  error: false,
  data: [],
};

const numOfUmis = {
  loading: true,
  error: false,
  data: [],
};

const initialState = {
  mitochondrialContent,
  doubletScores,
  numOfGenes,
  numOfUmis,
};

export default initialState;
