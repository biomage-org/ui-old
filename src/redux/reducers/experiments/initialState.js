const experimentTemplate = {
  name: null,
  description: 'Analysis description',
  id: null,
  createdAt: null,
  updatedAt: null,
  notifyByEmail: true,
  meta: { organism: null, type: '10x' },
  sampleIds: [],
  metadataKeys: [],
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
    activeExperimentId: null,
  },
  ids: [],
};

const METADATA_DEFAULT_VALUE = 'N.A.';

export default initialState;
export { experimentTemplate, METADATA_DEFAULT_VALUE };
