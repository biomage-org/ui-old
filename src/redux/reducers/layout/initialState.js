const initialState = {
  windows: {
    direction: 'row',
    first: {
      first: {
        first: 'UMAP', second: 'Cell sets and Metadata', direction: 'row', splitPercentage: 60,
      },
      second: 'Heatmap',
      direction: 'column',
      splitPercentage: 50,
    },
    second: 'Genes',
    splitPercentage: 70,
  },
  panel: 'Gene list',
};

export default initialState;
