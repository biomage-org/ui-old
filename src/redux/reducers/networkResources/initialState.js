const initialState = {
  environment: undefined,
  domainName: undefined,
};

export default initialState;
