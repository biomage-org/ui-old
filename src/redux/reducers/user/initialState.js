const initialState = {
  current: null,
};

export default initialState;
