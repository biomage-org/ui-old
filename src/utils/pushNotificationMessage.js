import { message } from 'antd';

const pushNotificationMessage = (type, text, duration = 4) => {
  switch (type) {
    case 'success':
      message.success(text, duration);
      break;
    case 'error':
      message.error(text, duration);
      break;
    case 'info':
      message.info(text, duration);
      break;
    case 'warning':
    case 'warn':
      message.warn(text, duration);
      break;
    case 'loading':
      message.loading(text, duration);
      break;
    default:
      message.info(text, duration);
      break;
  }
};

export default pushNotificationMessage;
