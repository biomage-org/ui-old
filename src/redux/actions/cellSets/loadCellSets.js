import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import {
  CELL_SETS_LOADED, CELL_SETS_LOADING, CELL_SETS_ERROR,
} from 'redux/actionTypes/cellSets';
import endUserMessages from 'utils/endUserMessages';

import config from 'config';
import { api } from 'utils/constants';

const loadCellSets = (experimentId, forceReload = false) => async (dispatch, getState) => {
  const {
    loading, error, updatingClustering,
  } = getState().cellSets;

  if (!forceReload && ((!loading && !error) || updatingClustering)) {
    return null;
  }

  if (error) {
    dispatch({
      type: CELL_SETS_LOADING,
    });
  }

  let url;

  if (config.currentApiVersion === api.V1) {
    url = `/v1/experiments/${experimentId}/cellSets`;
  } else if (config.currentApiVersion === api.V2) {
    url = `/v2/experiments/${experimentId}/cellSets`;
  }

  try {
    const data = await fetchAPI(url);

    dispatch({
      type: CELL_SETS_LOADED,
      payload: {
        experimentId,
        data: data.cellSets,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_CELL_SETS);

    dispatch({
      type: CELL_SETS_ERROR,
      payload: { error: errorMessage },
    });
  }
};

export default loadCellSets;
