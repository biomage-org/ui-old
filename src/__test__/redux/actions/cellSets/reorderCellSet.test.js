import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import waitForActions from 'redux-mock-store-await-actions';
import { waitFor } from '@testing-library/react';

import reorderCellSet from 'redux/actions/cellSets/reorderCellSet';
import { CELL_SETS_REORDER } from 'redux/actionTypes/cellSets';
import pushNotificationMessage from 'utils/pushNotificationMessage';

import initialState from 'redux/reducers/cellSets/initialState';

enableFetchMocks();

jest.mock('config');

const mockStore = configureStore([thunk]);

describe('reorderCellSet action', () => {
  const experimentId = '1234';
  const rootKey = 'root';
  const cellSetKey = 'child';

  const cellSetsNodeState = {
    ...initialState,
    properties: {
      [rootKey]: {
        rootNode: true,
      },
      [cellSetKey]: {
        rootNode: false,
        parentNodeKey: rootKey,
      },
    },
  };

  const response = new Response(JSON.stringify({}));

  beforeEach(() => {
    jest.resetAllMocks();

    fetchMock.resetMocks();
    fetchMock.doMock();
  });

  it('Works correctly', async () => {
    fetchMock.mockResolvedValueOnce(response);

    const store = mockStore({ cellSets: { ...cellSetsNodeState, loading: false } });
    await store.dispatch(reorderCellSet(experimentId, cellSetKey, 5));

    // CELL_SETS_REORDER action is sent
    await waitForActions(store, [CELL_SETS_REORDER]);
    expect(store.getActions()[0]).toMatchSnapshot();

    // Fetch was called correctly
    expect(fetchMock).toHaveBeenCalledTimes(1);

    const [url, body] = fetchMock.mock.calls[0];
    expect(url).toEqual('http://localhost:3000/v2/experiments/1234/cellSets');
    expect(body).toMatchSnapshot();
  });

  it('Does not dispatch the reorder action on fetch error', async () => {
    fetchMock.mockRejectOnce(() => Promise.reject(new Error('Some error')));

    const store = mockStore({ cellSets: { ...cellSetsNodeState, loading: false } });
    await store.dispatch(reorderCellSet(experimentId, cellSetKey, 5));

    // ERROR notification is shown
    await waitFor(() => expect(pushNotificationMessage).toHaveBeenCalledTimes(1));
    const pushNotificationMessageParams = pushNotificationMessage.mock.calls[0];

    expect(pushNotificationMessageParams).toEqual(['error', 'We couldn\'t save your data.']);

    expect(store.getActions()).toHaveLength(0);
  });

  it('Uses V2 URL when using API version V2', async () => {
    const store = mockStore({ cellSets: { ...cellSetsNodeState, loading: false } });
    await store.dispatch(reorderCellSet(experimentId, cellSetKey, 5));

    expect(fetchMock).toHaveBeenCalledTimes(1);

    const [url, body] = fetch.mock.calls[0];

    expect(url).toEqual('http://localhost:3000/v2/experiments/1234/cellSets');
    expect(body).toMatchSnapshot();
  });
});
