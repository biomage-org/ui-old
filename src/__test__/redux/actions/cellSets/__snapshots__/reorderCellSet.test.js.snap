// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reorderCellSet action Uses V2 URL when using API version V2 1`] = `
Object {
  "body": "[{\\"$match\\":{\\"query\\":\\"$[?(@.key == \\\\\\"root\\\\\\")]\\",\\"value\\":{\\"children\\":[{\\"$match\\":{\\"query\\":\\"$[?(@.key == \\\\\\"child\\\\\\")]\\",\\"value\\":{\\"$move\\":5}}}]}}}]",
  "headers": Object {
    "Content-Type": "application/boschni-json-merger+json",
  },
  "method": "PATCH",
}
`;

exports[`reorderCellSet action Works correctly 1`] = `
Object {
  "payload": Object {
    "cellClassKey": "root",
    "cellSetKey": "child",
    "newPosition": 5,
  },
  "type": "cellSets/reorder",
}
`;

exports[`reorderCellSet action Works correctly 2`] = `
Object {
  "body": "[{\\"$match\\":{\\"query\\":\\"$[?(@.key == \\\\\\"root\\\\\\")]\\",\\"value\\":{\\"children\\":[{\\"$match\\":{\\"query\\":\\"$[?(@.key == \\\\\\"child\\\\\\")]\\",\\"value\\":{\\"$move\\":5}}}]}}}]",
  "headers": Object {
    "Content-Type": "application/boschni-json-merger+json",
  },
  "method": "PATCH",
}
`;
